### PR TITLE
mold: wrap with ld-wrapper

### PR DESCRIPTION
**Pull request by tobim**
###### Description of changes

Wrapping mold so it can find libraries from the nix build environment.

See #24744 for additional context. The situation with lld is very similar, but mold was easier to fix.

Demo:
```
$ printf '#include <iostream>\nint main(){std::cout << "Woo!\\\n";}\n' \
  | c++ -B$(nix build --print-out-paths .#mold)/bin -x c++ - \
  && readelf -p .comment a.out \
  && ./a.out

String dump of section '.comment':
  [     0]  GCC: (GNU) 10.3.0
  [    12]  mold 1.2.1 (compatible with GNU ld)
  [    37]  GCC: (GNU) 8.3.0

Woo!
```

- [ ] check in a `nix-shell` with additional dependencies
- [ ] verify that this doesn't involuntarily break linking when used with rust

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cmm** [comment]
would be totally neat if this got merged

**Comment by a-kenji** [comment]
I tested the following under `x86_64-linux`:

- check in a nix-shell with additional dependencies
- verify that this doesn't involuntarily break linking when used with rust: I can't verify that, but I tested it with various projects that link with shared libraries (sqlite3, openssl, gtk, adwaita) and it worked flawlessly in my tests

File: pkgs/development/tools/mold/default.nix [hunk 1]
```diff
@@ -30,6 +31,35 @@ stdenv.mkDerivation rec {
   LTO = 1;
   makeFlags = [ "PREFIX=${placeholder "out"}" ];
 
+  suffixSalt = lib.replaceStrings ["-" "."] ["_" "_"] stdenv.targetPlatform.config;
+  wrapperName = "MOLD_WRAPPER";
+  coreutils_bin = lib.getBin coreutils;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:36]: Where are those three used?

File: pkgs/development/tools/mold/default.nix [hunk 2]
```diff
@@ -30,6 +31,35 @@ stdenv.mkDerivation rec {
...
+  postInstall = let
+    targetPrefix = lib.optionalString (stdenv.targetPlatform != stdenv.hostPlatform)
+                                          (stdenv.targetPlatform.config + "-");
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:39]:
```suggestion
    targetPrefix = lib.optionalString (stdenv.targetPlatform != stdenv.hostPlatform) (stdenv.targetPlatform.config + "-");
```

File: pkgs/development/tools/mold/default.nix [hunk 3]
```diff
@@ -30,6 +31,35 @@ stdenv.mkDerivation rec {
...
+  in ''
+    mkdir -p $out/nix-support
+
+    mv $out/bin/mold $out/bin/.mold
+
+    export prog=$out/bin/.mold
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:45]: Why not use wrapProgram or makeWrapper?

Comment (tobim) [reply pkgs/development/tools/mold/default.nix:45]: i don't understand how those tools would help solve the problem in a better way.

File: pkgs/development/tools/mold/default.nix (hunk 1 above)
Comment (tobim) [reply pkgs/development/tools/mold/default.nix:36]: They are referenced in templates that get instantiated with the calls to `substituteAll` below. We could just `export` them right in the postInstall script, but defining them in nix code reduces volatility.

i'll add a comment.

File: pkgs/development/tools/mold/default.nix (hunk 3 above)
Comment (Mic92) [reply pkgs/development/tools/mold/default.nix:49]: I don't think this would help here.

**Comment by sternenseemann** [comment]
Linkers in nixpkgs are left unwrapped intentionally – our wrapper scripts are quite invasive and we want users to be able to deal with more low-level situations manually as well – outside of `stdenv` the wrapper scripts are useless anyways.

Therefore, you work should be split into `mold` and `mold-unwrapped` or similar. Additionally, the wrapped version should probably utilise `wrapBintools` (not sure where to pull the other bintools from best, maybe llvm?), so it can be used in a custom stdenv.

**Comment by RaitoBezarius** [comment]
Can I help to get this merged?

**Comment by tobim** [comment]
I don't have any capacity to work on this at the moment. Fell free to open a new PR and work in the suggestions from @sternenseemann.

**Comment by lf-** [comment]
@RaitoBezarius I've poked at using wrapBintools for this for a proof of concept for work. One big issue with it is that it's less than helpful on macOS without some patching. Further, it's hard to test since changing the original file will be a world rebuild. We should possibly introduce a new parameter to it or use a postInstall hook or something to work around this issue.

```nix
{ mold
, stdenv
, libcCross
, callPackage
, darwin
}:
let wrapBintoolsWith2 =
  { bintools
  , libc ? if stdenv.targetPlatform != stdenv.hostPlatform then libcCross else stdenv.cc.libc
  , ...
  } @ extraArgs:
  callPackage ./pkgs/build-support/bintools-wrapper2 (
    let self = {
      nativeTools = stdenv.targetPlatform == stdenv.hostPlatform && stdenv.cc.nativeTools or false;
      nativeLibc = stdenv.targetPlatform == stdenv.hostPlatform && stdenv.cc.nativeLibc or false;
      nativePrefix = stdenv.cc.nativePrefix or "";

      noLibc = (self.libc == null);

      inherit bintools libc;
      inherit (darwin) postLinkSignHook signingUtils;
    } // extraArgs; in self
  );
in
wrapBintoolsWith2 {
  bintools = mold;
}
```

I've patched binutils-wrapper2/default.nix like so:

```
      # added ld.mold ld64.mold
      for variant in ld.gold ld.bfd ld.lld ld.mold ld64.mold; do
        local underlying=$ldPath/${targetPrefix}$variant
        [[ -e "$underlying" ]] || continue
        wrap ${targetPrefix}$variant ${./ld-wrapper.sh} $underlying
      done
```

**Comment by 9999years** [comment]
Anything blocking this?

**Comment by Artturin** [comment]
adding mold to the variant list https://github.com/NixOS/nixpkgs/pull/216383


@lf- postInstall isn't being run because you're overriding installPhase and not using `runHook postInstall`

**Comment by Artturin** [comment]
i think https://github.com/NixOS/nixpkgs/pull/216383 replaces the need for this pr

**Pull request closed without merging**